### ISSUE TITLE
Lecture 5. Backpropagation

### ISSUE THREAD

**Issue by seonghunn** [seonghunn/CS182#5]
# Part 1
linear regression model의 MSE loss의 graph 표현

<img width="322" alt="image" src="https://user-images.githubusercontent.com/69765278/159865146-47ceca0a-f7d4-4740-a154-582a8da015bb.png">

<img width="292" alt="image" src="https://user-images.githubusercontent.com/69765278/159865190-be1a0ca6-9f4f-47d8-95c4-b58a5568a205.png">

NN에서는 벡터를 자주 사용하므로, graph의 요소들을 vector로 취급한다.
위의 표현을 vector를 사용하면

<img width="271" alt="image" src="https://user-images.githubusercontent.com/69765278/159865321-fe4647b9-760a-493b-b66c-b6a70966a289.png">

위와 같이 표현할 수 있다.

## Logisic regression
Logistic regression에서의 p와 NLL

<img width="280" alt="image" src="https://user-images.githubusercontent.com/69765278/159865472-fd4de7b2-cee1-41c6-a48f-cfab6778c6b3.png">

이 NLL을 graph로 나타내 보자.

<img width="350" alt="image" src="https://user-images.githubusercontent.com/69765278/159865659-371d40ed-c91a-4f5a-b0de-928134ea5a87.png">

이 때 y는 one hot vector로, 원하는 값 이외에는 전부 0인 vector이다.
위의 복잡한 graph를 개선하기 위해, 다음과 같이 matrix를 사용한다

<img width="591" alt="image" src="https://user-images.githubusercontent.com/69765278/159866175-4eaa7087-faf2-46f0-92df-25a3f97c62fb.png">

<img width="654" alt="image" src="https://user-images.githubusercontent.com/69765278/159871173-4db3be35-4c98-40d1-9f17-fb1e8d1cae95.png">

variable에는 두 종류가 있는데 하나는 data, 하나는 parameters 이다.
이 때 data x, y는 input과 target output을 의미하며
&theta;가 의미하는 param은 한 특정 연산에 관여한다(CNN에서는 param share가 일어남)
이를 정리하면 아래 그림과 같다.

<img width="674" alt="image" src="https://user-images.githubusercontent.com/69765278/159871662-c9c711f8-562f-439f-95e8-2ce343745c24.png">

여기서 linear layer은 matrix vector product 연산을 하는 구간이고, fully connected layer라고도 한다.
그리고 cross entropy loss는 y가 없으면 구할 수 없다. 따라서 필수적으로 y에 의존한다.
위의 그림을 통해 x라는 input이 들어와 &theta; 라는 parameter와 곱해지는 linear layer을 거쳐 softmax를 취한 뒤 y와의 cross entropy loss를 줄이는 방식으로 training 된다는 것을 알 수 있다.

## Neural Network diagrams

<img width="328" alt="image" src="https://user-images.githubusercontent.com/69765278/159872157-371498fa-341f-4cb3-a181-c0968dd6aa44.png">

z<sup>(1)</sup>은 linear layer의 결과물이며, softmax는 parameter를 따로 가지지 않는 함수이기 때문에 다른 색으로 표현하였다.

## Logistic regression with features
잘 정리된 데이터라면 softmax를 취하면 된다.
하지만 복잡하게 분포한 데이터라면 softmax 내에 특정 함수를 설정하여 feature를 알아내야 한다

<img width="492" alt="image" src="https://user-images.githubusercontent.com/69765278/159873272-337209ae-67db-4d5d-a67e-035f2dec4005.png">

여기서 &theta; 의 dimensionality는 class가 2개 있고 우리가 설정한 함수 &phi;(x) 가 5 by 1이므로, 2 by 5 matrix가 될 것이다. 만약 bias term이 있다면 2 by 6 가 될 것이다.

<img width="630" alt="image" src="https://user-images.githubusercontent.com/69765278/159874171-8af3911e-ad88-4b1e-8582-d3222f830c6e.png">

이 때, 각각의 층에서 많은 parameter가 있기 때문에 이를 w로 표현한다. &theta;는 parameter 전체를 표현하기 위해 사용할 것.
이 때 각각의 feature가 **binary** logistic regression의 output이라면 위와 같이 softmax를 씌웠을 때 각각의 feature에 대해 sigmoid 형태로 표현할 수 있다 (binary output일 때는 분모 분자를 약분해 sigmoid 형태로 softmax를 단순화 시킬 수 있다)
모든 개별적인 softmax를 쌓아 &sigma(W<sup>(1)</sup>x) 로 정리할 수 있는데  이는 softmax와 다르다.
**각각의 독립적인 원소** 에서 feature를 뽑아내기 위해 sigmoid를 취한 것이므로, softmax와는 다르다는 것에 유의해야 한다.
이를 아래와 같이 표현할 수 있다

<img width="507" alt="image" src="https://user-images.githubusercontent.com/69765278/159876637-62c1f8bc-c182-471c-aa12-d014eb4c8c97.png">

여기서 activation function인 sigmoid는 parameter가 없다. 그 결과를 다른 linear layer를 통과시킨 후 softmax를 거쳐 Cross entropy loss를 구할 수 있다.

NN 버전으로 표현하면 다음과 같다.

<img width="656" alt="image" src="https://user-images.githubusercontent.com/69765278/159877021-07180624-f2b0-4e66-bd6d-dd36cc5450ff.png">

<img width="679" alt="image" src="https://user-images.githubusercontent.com/69765278/159877423-8eabca47-1a37-492b-93fa-24f7fc2b330c.png">

위와 같은 방식으로 layer을 많이 통과하도록 하는 학습법을 deep learning이라고 한다

## Activation function
sigmoid 혹은 softmax를 쓸 필요 없이 **non-linear function**을 사용할 수 있다
non linear인 이유는 weight sum 혹은 linear layer를 비선형 a 없이 계속 통과하게 된다면 단지 matrix의 선형 곱이 결과물이 되므로 이를 하나의 matrix와의 선형 연산으로 표현할 수 있게 된다.
따라서 **non-linear function**을 사용해야 하는 것이다.
충분한 층이 있고 nonlinear activation function을 사용하게 되면 거의 모든 것을 표현할 수 있다.

## what's so neural about it?
뉴런은 입력을 받아서 출력을 보낼지 말지를 결정하게 된다. 그 출력이 아래 뉴런으로 전달되게 된다.
인공 뉴런은 상위 계층의 뉴런의 weight sum을 받는다. 그리고 얼마만큼의 출력을 보내야 할지를 weight sum의 결과를 통해 판단하게 되고, activation function을 통해 이를 결정해 다음 Layer의 인공 뉴런으로 보내게 된다.

# Part 2
## Training neural networks
1. model class를 정의
2. loss function을 정의
3. optimizer 결정

<img width="143" alt="image" src="https://user-images.githubusercontent.com/69765278/160030584-65bb1baf-ef61-4277-9c82-3455bdc46766.png">

optimizer는 stochastic gradient descent가 주로 이용되는데, gradient descent는 각각의 parameter에 대한 미분값을 지닌 matrix이다. 따라서 우리가 model을 변경하는 일이 생기게 되면 gradient vector를 일일히 다시 계산하는 데에는 오랜 시간이 걸린다. 따라서 Neural Net의 gradient 연산을 자동화하는 algorithm이 필요하고 이것이 **backpropagation**이다.

## chain rule
우리가 일변수 함수, 스칼라 함수에서 chain rule을 쓰는 것처럼 matrix에서도 chain rule을 사용할 수 있다.

<img width="386" alt="image" src="https://user-images.githubusercontent.com/69765278/160030957-9050332b-df33-4127-a134-7a88b181bf63.png">

이 때 연산결과를 1 by m row matrix와 m by 1 col matrix의 dot product로 표현할 수 있다.
이를 n개의 x<sub>i</sub> 에 대해 일반화 하여 각각의 xi를 row에 쌓아 dy/dx 라는 n by m matrix로 일반화 하여 연산할 수 있다.

아래는 위의 High-dimensional chain rule의 이해를 돕기 위한 수식이다.

<img width="501" alt="image" src="https://user-images.githubusercontent.com/69765278/160031376-7ce549c3-8ef9-40c5-ba54-49c1914ce6a3.png">

예시로, Implicit differentiation을 들어보면 다음과 같다

<img width="451" alt="image" src="https://user-images.githubusercontent.com/69765278/160031488-bc9997a5-babb-4a82-a1c3-79d7fd2f8bdf.png">

출처: https://blog.naver.com/PostView.nhn?blogId=gsurb321&logNo=221629101321

위의 연산을 수행할 때 두가지 convention이 있다.

<img width="566" alt="image" src="https://user-images.githubusercontent.com/69765278/160032202-4a6408b1-6f92-4a3d-8663-337c726df69b.png">

이 강의에서의 convention에 따르면 z=f(y) 에서 y가 m dim col vector이면 dz/dy가 m dim col vector이다. 그리고 dy/dx는 n x m dim이다. (m개의 y, n개의 x)
다른 convention에서는 dz/dy 를 m dim row vector로 설정한다. 이렇게 되면 chain rule의 연산 순서가 뒤집힌다. 이 row vector convention이 자주 사용되는 이유는 linear function의 derivative를 구하게 되면(matrix multplication의 결과 등등) input과 derivative의 결과를 매칭시켜주기 때문이다. 이 강의에서의 convention은 이 결과의 transpose가 나온다.

이를 NN에 적용해 보자

<img width="489" alt="image" src="https://user-images.githubusercontent.com/69765278/160033400-eef22e4f-04c6-4d00-b8bb-96be6e8bcdc1.png">

우리는 이 NN에서 chain rule을 다음과 같이 사용할 수 있다.

<img width="222" alt="image" src="https://user-images.githubusercontent.com/69765278/160033773-fffd18db-2ed9-467a-9417-af6f2b1af026.png">

<img width="137" alt="image" src="https://user-images.githubusercontent.com/69765278/160033716-a99021e9-5216-4027-9d98-f102d8dfacf8.png">

우리는 row vector convention을 사용하고 있지 않기 때문에 W<sup>(2)</sup> 의 transpose가 나오게 된다.
이렇게 계산하는 것이 나쁜 이유는 위의 연산을 하는 데에 시간이 너무 오래 걸리기 때문이다.
각각의 z와 a가 n dim을 가진다면, 각각의 Jacobian은 n by n dim을 가지고, matrix multiplication에 O(n<sup>3</sup>)의 시간이 든다. 시간복잡도 측면에서 너무 나쁘다.

이를 효과적으로 하기 위해서 다음과 같은 사실을 이용한다.

<img width="517" alt="image" src="https://user-images.githubusercontent.com/69765278/160034492-e1860777-83b0-4e7d-bc4f-4308881ca4b5.png">

중간 연산에서의 J는 n by n 이지만, 마지막 J는 항상 n by 1이다. 이는 loss가 scalar valued이기 때문에 항상 사실이다. n x n 과 n x 1 은 곱하는 데에 O(n<sup>2</sup>)의 시간이 소요되어 O(n<sup>3</sup>) 보다 효율적이기 때문에, 가장 오른쪽부터 연산을 시작한다. 오른쪽의 연산 결과를 &delta; 라고 하면 이는 n x 1 matrix이기 떄문에 다시 바로 앞의 n x n J matrix와 곱하면 O(n<sup>2</sup>)의 시간복잡도로 연산을 수행할 수 있다.
이렇게 오른쪽에서 시작해서 연산을 수행한다면 한 연산당 시간복잡도 O(n<sup>2</sup>) 에 연산을 수행 가능하다.

이를 예시를 들어 확인해 보도록 하자

<img width="621" alt="image" src="https://user-images.githubusercontent.com/69765278/160035390-8321b400-2a56-4807-ae6e-4a3b25ef1e72.png">

**&delta; 의 초기값, f(output) ,x<sub>f</sub>(input)와 &theta;<sub>f</sub>(parameter, W<sup>(n)</sup> in linear layer) 가 무얼 의미하는지를 확인하자.** 이렇게 &delta 와 df/d&theta<sub>f</sub>와의 연산을 갱신하며 계산하는 것이 backpropagation이다.

<img width="682" alt="image" src="https://user-images.githubusercontent.com/69765278/160035625-0db8d28e-fa2b-4b80-bb2a-00fbbaa66a17.png">

모든 &theta;에 대해 dL/d&theta; 를 구해야 하기 때문에 W(1), W(2) 에 대한 미분값을 전부 구해야 한다. 
sigmoid에서의 계산에서는 parameter가 없기 때문에 &theta<sub>f</sub>가 존재하지 않고 따라서 df/dx<sub>f</sub> 를 계산한다. 이는 output/input에 해당하는 da<sup>(1)</sup>/dz<sup>(1)</sup> 가 된다. 이렇게 맨 앞에 갈 때까지 연산을 수행하는 방식으로 backpropagation을 수행한다.

# Part 3

<img width="500" alt="image" src="https://user-images.githubusercontent.com/69765278/160036140-9fbc676b-067c-4254-9596-78cc5929c8c8.png">

여기서 ReLU는 layer가 클 때도 수행 성능이 좋으며, derivative가 굉장히 단순해 사용하기 쉽다.
추가적으로 derivative vanishing 현상도 방지할 수 있다.

## Bias terms
a<sup>(i)</sup>가 0이라면 우리는 어떤 연산에서도 답이 0이 나올수밖에 없다. 따라서 우리는 단순히 W와 a의 곱으로 output을 보내는 것이 아니고 bias term을 다음과 같이 추가한다.

<img width="485" alt="image" src="https://user-images.githubusercontent.com/69765278/160036414-94b35777-1347-404b-905c-935e799799b0.png">

이렇게 bias term을 추가하게 되면, parameter는 W와 b 두 개가 된다.

## Backporpagtion의 linear layer에서의 연산

<img width="659" alt="image" src="https://user-images.githubusercontent.com/69765278/160037180-3cbb1baf-67a2-463f-8a33-1d4969a02b96.png">

dz/dW 와 &delta; 의 곱은 위의 그림과 같이 표현되며 z<sub>i</sub>는 W<sub>ik</sub> 에만 변수로서 관련 있기 때문에 (3차원 큐브 모양을 생각해 보거나, f(x,y)에서 x로 미분하면 y는 상수 취급이라 사라지는것을 생각하면 된다) 나머지에 대한 미분값은 전부 0이고, j=i일 때만 결과값을 가지게 되는데 위에서 보듯이 zi는 matrix multiplication으로 정의되어 있기 때문에 chain rule을 통해 결과는 a<sub>k</sub> 가 된다.

따라서 dz3/dW 의 예시를 보면, 3rd row를 제외하고는 전부 0이된다. 따라서 &delta; 또한 3에 해당하는 것 제외하고는 0이 되기 때문에

<img width="165" alt="image" src="https://user-images.githubusercontent.com/69765278/160037854-4567499c-686c-4806-866c-a26990343a54.png">

위와 같은 결과가 나오고

<img width="178" alt="image" src="https://user-images.githubusercontent.com/69765278/160037904-919ed3d7-0968-4ecb-ab28-685e3dc44cad.png">

위 matrix를 outer product로 나타낼 수 있기 때문에 결과가 &delta;a<sup>T</sup> 로 나오는 것을 알 수 있다.

## b에 대한 연산

<img width="486" alt="image" src="https://user-images.githubusercontent.com/69765278/160038316-e065ee29-11ec-4367-98cb-c259f9325afa.png">

b 또한 bi만이 zi에 대한 변수로 작용하기 때문에, zi 와 bj에서 i와 j가 같을 때만 값이 나오게 된다. 하지만 상수 term의 미분값은 1이기 때문에 dz/db 를 정리하면 Identity Matrix I가 되게 된다.

## a(input)에 대한 연산

<img width="486" alt="image" src="https://user-images.githubusercontent.com/69765278/160038444-6ee1c443-d56a-47f0-bab0-8c941146a5ec.png">

chain rule에 의해 dzi/dak는 위와 같이 W<sub>ik</sub>로 표현되고, 이를 정리하면 W의 transpose matrix가 결과로 나오게 된다.

## 결과

<img width="486" alt="image" src="https://user-images.githubusercontent.com/69765278/160038621-b65c9361-23e7-4e37-9484-2adda3df69cb.png">

## sigmoid에서의 연산
parameter가 없는 연산이고, input에 해당하는 부분만 output이 존재하기 때문에 아래와 같이 diagonal matrix로 표현하게 된다.

<img width="165" alt="image" src="https://user-images.githubusercontent.com/69765278/160038756-f9134e84-fd5c-44ca-9760-6f38e7c64746.png">

연산 결과는 아래와 같다. 이 단계에서는 df/dz 를 구하게 된다

<img width="548" alt="image" src="https://user-images.githubusercontent.com/69765278/160038902-f692f32b-33b5-4cf9-9eb9-0f02d2312cf4.png">

## ReLU에서의 연산

<img width="311" alt="image" src="https://user-images.githubusercontent.com/69765278/160039005-54e8a278-eb80-4dfc-817a-2f28a58ba6ce.png">

## Summary

Backpropagation이란, &delta;를 뒤에서부터 갱신하며 더 좋은 시간 복잡도로 연산을 수행해 내는 것이다. 그 과정에서 각 단계마다 구해야 하는 df/dW, df/dx, df/dz 등등을 구하며 &delta;를 갱신하며 gradient를 구하게 된다.

<img width="567" alt="image" src="https://user-images.githubusercontent.com/69765278/160039156-e7ff9e69-cfbe-42d8-80b7-0b9c495c97d1.png">
